// rettiwt-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataValidationError = exports.Query = exports.Variables = exports.TweetFilter = exports.Features = exports.Args = exports.Url = exports.EErrors = exports.EResourceType = void 0;
// Exporting enums
var Resources_1 = require("./enums/Resources");
Object.defineProperty(exports, "EResourceType", { enumerable: true, get: function () { return Resources_1.EResourceType; } });
var Errors_1 = require("./enums/Errors");
Object.defineProperty(exports, "EErrors", { enumerable: true, get: function () { return Errors_1.EErrors; } });
// Exporting classes
var Url_1 = require("./Url");
Object.defineProperty(exports, "Url", { enumerable: true, get: function () { return Url_1.Url; } });
// Exporting custom types
__exportStar(require("./types/request/params/Query"), exports);
// Exporting base raw types
__exportStar(require("./types/response/base/DataResult"), exports);
__exportStar(require("./types/response/base/Tweet"), exports);
__exportStar(require("./types/response/base/User"), exports);
// Exporting payload models
var Args_1 = require("./models/payloads/Args");
Object.defineProperty(exports, "Args", { enumerable: true, get: function () { return Args_1.Args; } });
var Features_1 = require("./models/payloads/Features");
Object.defineProperty(exports, "Features", { enumerable: true, get: function () { return Features_1.Features; } });
var TweetFilter_1 = require("./models/payloads/TweetFilter");
Object.defineProperty(exports, "TweetFilter", { enumerable: true, get: function () { return TweetFilter_1.TweetFilter; } });
var Variables_1 = require("./models/payloads/Variables");
Object.defineProperty(exports, "Variables", { enumerable: true, get: function () { return Variables_1.Variables; } });
// Exporting param models
var Query_1 = require("./models/params/Query");
Object.defineProperty(exports, "Query", { enumerable: true, get: function () { return Query_1.Query; } });
// Exporting other models
var DataValidationError_1 = require("./models/errors/DataValidationError");
Object.defineProperty(exports, "DataValidationError", { enumerable: true, get: function () { return DataValidationError_1.DataValidationError; } });
