// rettiwt-api/dist/graphql/enums/Errors.js
"use strict";
exports.__esModule = true;
exports.DataErrors = exports.ValidationErrors = void 0;
/**
 * Different types error messages for validation errors returned by GraphQL API.
 *
 * @public
 */
var ValidationErrors;
(function (ValidationErrors) {
    ValidationErrors["NoUserIdentification"] = "Either userName or id must be given";
})(ValidationErrors = exports.ValidationErrors || (exports.ValidationErrors = {}));
;
/**
 * Different type of error messages related to data that are returned by GraphQL API.
 *
 * @public
 */
var DataErrors;
(function (DataErrors) {
    DataErrors["NoTweetsFound"] = "No tweets matching the given criteria found";
    DataErrors["NoLikersFound"] = "No likers found for the tweet with the given id";
    DataErrors["NoRetweetersFound"] = "No retweeters found for the tweet with the given id";
    DataErrors["NoUserTweetsFound"] = "No tweets were found for the user with the given id";
    DataErrors["NoFollowsFound"] = "No follow details were found for the user with the given id";
    DataErrors["NoLikedTweetsFound"] = "No liked tweets were found for the user with the given id";
})(DataErrors = exports.DataErrors || (exports.DataErrors = {}));
;
