// rettiwt-api/dist/models/auth/AuthCookie.js
"use strict";
exports.__esModule = true;
exports.AuthCookie = void 0;
/**
 * The cookie containing the tokens that are used to authenticate against Twitter.
 *
 * @internal
 */
var AuthCookie = /** @class */ (function () {
    function AuthCookie(cookie) {
        var _a, _b, _c, _d;
        this.auth_token = (_a = cookie === null || cookie === void 0 ? void 0 : cookie.auth_token) !== null && _a !== void 0 ? _a : '';
        this.ct0 = (_b = cookie === null || cookie === void 0 ? void 0 : cookie.ct0) !== null && _b !== void 0 ? _b : '';
        this.kdt = (_c = cookie === null || cookie === void 0 ? void 0 : cookie.kdt) !== null && _c !== void 0 ? _c : '';
        this.twid = (_d = cookie === null || cookie === void 0 ? void 0 : cookie.twid) !== null && _d !== void 0 ? _d : '';
    }
    /**
     * @returns The string respresentation of this cookie in the valid cookie string format.
     */
    AuthCookie.prototype.toString = function () {
        /** The string representation of this cookie. */
        var cookieString = '';
        // Iterating through the (key, value) pairs of this cookie
        for (var _i = 0, _a = Object.entries(this); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            cookieString += "".concat(key, "=").concat(value, ";");
        }
        return cookieString;
    };
    return AuthCookie;
}());
exports.AuthCookie = AuthCookie;
